visionmetrics/detection/confusion_matrix.py

```diff
@@ -0,0 +1,171 @@
+            'FP_due_to_extra_pred_boxes': self.fp_due_to_extra_pred_boxes.item()
+        }
+
+    def _update_confusion_matrix(self, predictions: List[List[List[float]]], targets: List[List[List[float]]], iou_threshold: float) -> None:
+        for preds, gts in zip(predictions, targets):
+            # Check empty predictions and targets
+            if self._is_empty(preds):
+                self.fn += len(gts)
+                continue
+
+            if self._is_empty(gts):
+                self.fp += len(preds)
+                self.fp_due_to_extra_pred_boxes += len(preds)
+                continue
+
+            # Count per image
+            gt_boxes_used = [False] * len(gts)  # To keep track of all matched GT boxes
+            preds = sorted(preds, key=lambda x: x[1], reverse=True)
+            for idx, pred in enumerate(preds):
+                # If all GT boxes for this image have been matched (TP)
+                # with corresponding pred boxes then all remaining pred boxes are FPs
+                if all(gt_boxes_used):
+                    self.fp += len(preds) - idx
+                    self.fp_due_to_extra_pred_boxes += len(preds) - idx
+                    break
+
+                # Otherwise, calculate IoU with all remaining GT boxes
+                pred_class_id, pred_box = pred[0], pred[2:]
+                best_same_class_iou = 0
+                best_same_class_gt_index = -1
+                best_diff_class_iou = 0
+
+                for gt_index, gt in enumerate(gts):
+                    if gt_boxes_used[gt_index]:  # Skip GT box if already used
+                        continue
+
+                    gt_class_id, gt_box = gt[0], gt[1:]
+
+                    # Calculate IoU and update best pred box IoU with same and different class GT boxes
+                    current_iou = self.iou(pred_box, gt_box)
+                    if pred_class_id == gt_class_id:
+                        if current_iou > best_same_class_iou:
+                            best_same_class_iou = current_iou
+                            best_same_class_gt_index = gt_index
+                    else:
+                        if current_iou > best_diff_class_iou:
+                            best_diff_class_iou = current_iou
+
+                # Update TP and FP counts
+                if best_same_class_gt_index != -1:  # GT box with same class found
+                    if best_same_class_iou >= iou_threshold:
+                        self.tp += 1
+                        gt_boxes_used[best_same_class_gt_index] = True  # Mark GT box as used
+                    else:
+                        self.fp += 1
+                        self.fp_due_to_low_iou += 1
+                else:
+                    if best_diff_class_iou >= iou_threshold:
+                        self.fp += 1
+                        self.fp_due_to_wrong_class += 1
+                    else:
+                        self.fp += 1
+                        self.fp_due_to_low_iou += 1
```

Comment:
For seek of detailed analysis, how about split this "fp_due_to_low_iou" into "fp_due_to_low_iou_and_wrong_class" (IOU>0),  and "fp_due_to_no_match"? (IOU==0)

Reply:
**There are 8 cases to consider:**
1. iou >= thresh + correct class
2. iou >= thresh + wrong class
3. 0 < iou < thresh + correct class
4. 0 < iou < thresh + wrong class
5. iou = 0 + correct class
6. iou = 0 + wrong class
7. no gt box left to match
8. no pred box left to match

**Currently,**
Case 1 -> TP
Case 2 -> FP due to wrong class
Case 3,4,5,6 -> FP due to low IOU
Case 7 -> FP due to extra pred
Case 8 -> FN

I think if we split "fp_due_to_low_iou" into "fp_due_to_low_iou_and_wrong_class" and "fp_due_to_no_match", then we also need atleast "fp_due_to_low_iou_and_correct_class". The no_match (iou=0) case could also be split into correct and wrong classes. Basically, consider each case as a separate category? I suspect this

I think no match is just a subset of low_iou. So we could include no_match into low_iou cases depending on wrong and correct classes. 

TLDR: I suggest we split "fp_due_to_low_iou" into "fp_*_wrong_class" and "fp_*_correct_class" (these will also include the iou=0 cases)

Thoughts?

Reply:
@cy-bai - thoughts?

Reply:
LGTM thanks!